co2mond/src/main.c

```diff
@@ -122,33 +166,237 @@ static void
 write_heartbeat()
 {
     char buf[VALUE_MAX];
-    snprintf(buf, VALUE_MAX, "%lld", (long long)time(0));
+    const time_t now = time(0);
+    snprintf(buf, VALUE_MAX, "%lld", (long long)now);
     write_value("heartbeat", buf);
+    state_lock();
+    co2mon.heatbeat = now;
+    state_unlock();
+}
+
+static int
+read_match_path(const int fd)
+{
+    // Prefix has no \r\n to support HTTP/1.0 requests with no headers.
+    const char *prefix = "GET /metrics HTTP/1.";
+    const int len = strlen(prefix);
+    for (int i = 0; i < len; ++i)
+    {
+        char next;
+        const int rv = read(fd, &next, 1);
+        if (rv == 0) // EOF
+        {
+            return -1;
+        }
+        if (rv == -1)
+        {
+            perror("read");
+            return -1;
+        }
+        if (rv != 1 || next != prefix[i]) // mismatch
+        {
+            return -1;
+        }
+    }
+    return 0;
+}
+
+static int
+read_find_crlfcrlf(const int fd)
+{
+    const char *substring = "\x0d\x0a\x0d\x0a";
+    const int len = strlen(substring);
+    // automata is tivial as `substring` has no repeated chars
+    for (int i = 0; i < len; ++i)
+    {
+        assert(i >= 0);
+        char next;
+        const int rv = read(fd, &next, 1);
+        if (rv == 0) // EOF
+        {
+            return -1;
+        }
+        if (rv == -1)
+        {
+            perror("read");
+            return -1;
+        }
+        if (rv != 1) // WAT?
+        {
+            return -1;
+        }
+        if (next != substring[i])
+        {
+            if (next == substring[0])
+            {
+                i = 0;
+            }
+            else
+            {
+                i = -1;
+            }
+        }
+    }
+    return 0;
+}
+
+static void*
+prometheus_thread(void *arg)
+{
+    const int listen_fd = (ssize_t)arg;
+    while (1) {
+        const int client_fd = accept(listen_fd, NULL, NULL);
+        const struct timeval maxdelay = { 5, 0 }; // 5 seconds, just like co2mon_read_data()
+        struct co2mon_state copy;
+        FILE* out = NULL;
+
+        if (client_fd == -1)
+        {
+            perror("accept");
+            continue;
+        }
+
+        if (setsockopt(client_fd, SOL_SOCKET, SO_RCVTIMEO, &maxdelay, sizeof(maxdelay)) != 0 ||
+            setsockopt(client_fd, SOL_SOCKET, SO_SNDTIMEO, &maxdelay, sizeof(maxdelay)) != 0)
+        {
+            perror("setsockopt");
+            goto cleanup;
+        }
+
+        if (read_match_path(client_fd) != 0 || read_find_crlfcrlf(client_fd) != 0)
+        {
```

Comment:
@darkk what do you think about sending "400 Bad Request" in this case?

Reply:
I have weak arguments for both options (besides vague mumble about RFC-compliance as I don't send `Date` anyway).

Pro: sending `400` is a better response than RST that may aid in catching some typo somewhere.
Contra: sending `400` to a "disconnected" `client_fd` may deliver SIGPIPE in some weird case (but it should be ignored for a daemon anyway).

Thanks for highlighting the matter, it made me think of SIGPIPE that should be ignored.

I'll add `400 Bad Request` together with SIGPIPE SIG_IGN.